### PR TITLE
make target optional in decision points

### LINKED ISSUE #2513: Make target optional
We discussed making the target optional so that it could be used only when it makes sense.

Further clarifications needed for the changes in the UI and the backend.

UI changes
- Add Decision Point Modal: Site would be default field, with button or arrow dropdown to add optional Target
- Decision Point table - remove target column, only show when target is defined in the Add Decision Point modal
- Data tab - remove Target when shown 

Backend changes
- Change types 
- Target may still be a required field in the DB, possibly replace with empty string when unused? May be easier than defining a rule for decision points that use targets and those that don't use targets
- Refactoring (we still have DPs called "Partition" in several places)

We need a design for "stacking" the site and target in the displays - like on the payload card. Site, with target in parentheses. Apply this design everywhere we display DP with site & target. Truncate or wrap?

Image

Image

### PR DESCRIPTION
**Pull request by bcb37**
make 'target' an optional addition in the decision point modal; add backend tests to verify NULL targets

### CODE REVIEW

File: packages/backend/src/api/controllers/validators/MarkExperimentValidator.v5.ts [hunk 1]
```diff
@@ -22,7 +22,8 @@ class Data {
   site: string;
 
   @IsString()
-  target: string;
+  @IsOptional()
+  target?: string | null;
```
[the remark this answers is not in the document]
Comment (bcb37) [reply packages/backend/src/api/controllers/validators/MarkExperimentValidator.v5.ts:26]: '' is not a "no target" representation

File: packages/backend/src/api/controllers/ExperimentClientController.v6.ts [hunk 2]
```diff
@@ -451,7 +451,7 @@ export class ExperimentClientController {
       experiment.data.assignedCondition?.conditionCode ?? null,
       request.logger,
       experiment.data.assignedCondition?.experimentId ?? null,
-      experiment.data.target,
+      experiment.data.target ?? null,
       experiment.uniquifier ? experiment.uniquifier : null,
```
[the remark this answers is not in the document]
Comment (bcb37) [reply packages/backend/src/api/controllers/ExperimentClientController.v6.ts:455]: We'll make target not required

File: packages/backend/src/api/controllers/ExperimentClientController.v5.ts [hunk 3]
```diff
@@ -481,7 +481,7 @@ export class ExperimentClientController {
       experiment.data.assignedCondition?.conditionCode ?? null,
       request.logger,
       experiment.data.assignedCondition?.experimentId ?? null,
-      experiment.data.target,
+      experiment.data.target ?? null,
       experiment.uniquifier ? experiment.uniquifier : null,
```
[the remark this answers is not in the document]
Comment (bcb37) [reply packages/backend/src/api/controllers/ExperimentClientController.v5.ts:485]: empty strings are not null, and they're disallowed in the UI

**Comment by danoswaltCL** [comment]
UI design Q's for target optional... cc: @zackcl @amurphy-cl

is it a requirement to make adding a target a thing that we want to hide behind an add button?  i get that we want to make it clear that it's optional, but is there any problem with simply removing the "required" validator and marking this as optional in the placeholder text? if we do want an add button, we ought to style it like other buttons. 

image

Currently the required validator still exists when added, so after selecting add target, there's no way to remove it.

image

Design view w/o the site column, things feel far apart:

image

Data tab conditions enrolled minus the target, the column is just removed, maybe fine but it's noticeable:

image

Payloads looks good I think though

image

**Comment by danoswaltCL** [comment]
i wonder if the terminology will a little confusing if we are using the word `site` and `decision-point` interchangeably, without a target they are basically mean the same thing, but with a target they don't....

**Comment by zackcl** [comment]
Experiment - Add Decision Point

I think making "Target" optional is more consistent with the existing design, and it should be sufficient to indicate that it's optional since the "Create" button becomes enabled as soon as the user types something into "Site."

For the Decision Points table, I think we should leave the Target column header as is. This makes it clearer which rows have only a site and which have both a site and a target. The same applies to the Conditions table in the Data tab.

**Comment by zackcl** [comment]
Or maybe we can convert all "Site" and "Target" columns into a single "Decision Point" column, and display values as "site; target". I think this would simplify the UI, as long as users understand that a decision point consists of a site and an optional target combined with a semicolon delimiter.

**Comment by danoswaltCL** [comment]
the type for `target` is inconsistent, I think it should almost always be truly optional (`target?: string`) and not `string | null`, especially in client lib or UI.

copilot says as much in various ways it appears, but it's also saying to "normalize" falsey values to null, which i disagree with, that's just pointless strictness to say `target: string | undefined | null` or coercing with `target ?? null` when null is not meaningfully different from `undefined` or other falsey values.

**Comment by danoswaltCL** [comment]
But the most important part works, it seems to do everything right I can see infsofar as running and enrolling users w/o targets involved 👍

**Comment by bcb37** [comment]
> the type for `target` is inconsistent, I think it should almost always be truly optional (`target?: string`) and not `string | null`, especially in client lib or UI.
> 
> copilot says as much in various ways it appears, but it's also saying to "normalize" falsey values to null, which i disagree with, that's just pointless strictness to say `target: string | undefined | null` or coercing with `target ?? null` when null is not meaningfully different from `undefined` or other falsey values.

@danoswaltCL  The problem is that it is in fact meaningfully different is some cases, since it's stored as `null` in the database. For that reason, we have to make sure it's either a string or `null` all the way through the workflow, or make sure that 'undefined' gets converted to `null` for database interactions.

**Comment by bcb37** [comment]
> Experiment - Add Decision Point
> I think making "Target" optional is more consistent with the existing design, and it should be sufficient to indicate that it's optional since the "Create" button becomes enabled as soon as the user types something into "Site."
> 
> For the Decision Points table, I think we should leave the Target column header as is. This makes it clearer which rows have only a site and which have both a site and a target. The same applies to the Conditions table in the Data tab.

@zackcl  Ok, so it seems like we want to back out all the frontend changes that were in the spec (and implemented here), and just make the 'target' field optional in the modal?

**Comment by zackcl** [comment]
Sorry, I thought this was an issue and not a PR when I added that comment.

I just checked the spec, which seems to have been written by April. I personally disagree with the proposed UI changes and would suggest making the Target field optional while keeping the Target columns even when no target is defined, as this would be more consistent and reliable. That said, I’m open to others’ thoughts, and we may want to check with April on this.

**Comment by danoswaltCL** [comment]
is there any reason "no-target" can't just be an empty string? and avoid `null` altogether?

**Comment by bcb37** [comment]
> is there any reason "no-target" can't just be an empty string? and avoid `null` altogether?

It could be anything, as long as we agree on what it is. `null` made the most sense to me for two reasons: 1) An empty string it still a string, so we have to make a distinction between that particular string and all others from the point of view of typing; and 2) If it's `null` we don't have to convert it from whatever we decided it would be to `null` for interactions with the database.

**Comment by danoswaltCL** [comment]
it just all stays the almost the same, that's the point, i've been looking at it, i'm pretty confident that if we keep the type to always be string and treat undefined as empty string, plunk empty string in database, there's barely any code to change and it would be safer than introducing a new `null` possibility to have to deal with.

**Comment by bcb37** [comment]
Closing due to change in spec.
